### PR TITLE
Make EventId on WeekViewEvent generic for custom calendar fetches

### PR DESCRIPTION
**Pull request by udaysrinath**
Fixes issue : https://github.com/Quivr/Android-Week-View/issues/44

Note : 
(1) I rebased and that pulled in a lot of other changes from upstream, hope thats fine take a peek. 
(2) Tested for both Basic and Async Activity on the sample app and the change looks good.

@jhoobergs Please Review

### CODE REVIEW

File: library/src/main/java/com/alamkanak/weekview/WeekViewEvent.java [hunk 1]
```diff
@@ -193,11 +216,6 @@ public boolean equals(Object o) {
 
     }
 
-    @Override
```
Comment (jhoobergs) [inline library/src/main/java/com/alamkanak/weekview/WeekViewEvent.java]: I think this method should remain. I guess you can return mId.hashCode() or something

File: library/src/main/java/com/alamkanak/weekview/WeekView.java [hunk 2]
```diff
@@ -301,7 +301,7 @@ public boolean onSingleTapConfirmed(MotionEvent e) {
                 List<EventRect> reversedEventRects = mEventRects;
                 Collections.reverse(reversedEventRects);
                 for (EventRect eventRect : reversedEventRects) {
-                    if (eventRect.event.getId() != mNewEventId &&eventRect.rectF != null && e.getX() > eventRect.rectF.left && e.getX() < eventRect.rectF.right && e.getY() > eventRect.rectF.top && e.getY() < eventRect.rectF.bottom) {
+                    if (!eventRect.event.getId().equals(mNewEventId) && eventRect.rectF != null && e.getX() > eventRect.rectF.left && e.getX() < eventRect.rectF.right && e.getY() > eventRect.rectF.top && e.getY() < eventRect.rectF.bottom) {
```
Comment (jhoobergs) [inline library/src/main/java/com/alamkanak/weekview/WeekView.java]: I would write mNewEventId.equals(eventRect.event.getId()) to avoid possible null-pointer exceptions on the id of an event (mNewEventId won't be null)

File: library/src/main/java/com/alamkanak/weekview/WeekViewEvent.java [hunk 3]
```diff
@@ -113,10 +126,20 @@ public WeekViewEvent(long id, String name, String location, Calendar startTime,
...
     }
 
+    /**
+     * Initializes the event for week view.
+     * @param id The id of the event.
+     * @param name Name of the event.
+     * @param startTime The time when the event starts.
+     * @param endTime The time when the event ends.
+     */
+    public WeekViewEvent(int id, String name, Calendar startTime, Calendar endTime) {
```
Comment (jhoobergs) [inline library/src/main/java/com/alamkanak/weekview/WeekViewEvent.java]: Why did you add this one specifically ? I don't see the logic behind this.

**Review by jhoobergs — CHANGES_REQUESTED** [review]
Some small comments.
Also: I think we should make it backward compatible, so don't change long return types to String, but create a new one for the string Id and let the old one use Integer.parse().

File: library/src/main/java/com/alamkanak/weekview/WeekViewEvent.java (hunk 3 above)
Comment (udaysrinath) [reply library/src/main/java/com/alamkanak/weekview/WeekViewEvent.java]: These are overrides for the ```BaseActivity``` in the sample. I guess the helpers can move to the sample app, and dont need to be here.

**Comment by udaysrinath** [comment]
> Also: I think we should make it backward compatible, so don't change long return types to String, but create a new one for the string Id and let the old one use Integer.parse().

Done!

**Comment by jhoobergs** [comment]
@udaysrinath 
The backward compatibility isn't complete yet: 
- The constructors with an int as id should all still work
- The xml attr with an int should still work (so create a seperate one for the new string attribute)

After this, the samples should still work without any changes. 
When that's the case, you can add extra examples that use a string as id

**Comment by udaysrinath** [comment]
Good catch @jhoobergs! Hopefully I have caught all the backward compat scenarios. Also, updated the sample to accept both types. Added ```@Deprecated``` annotation explicitly.

File: library/src/main/java/com/alamkanak/weekview/WeekView.java [hunk 4]
```diff
@@ -442,6 +443,7 @@ public WeekView(Context context, AttributeSet attrs, int defStyleAttr) {
             mNewEventColor = a.getColor(R.styleable.WeekView_newEventColor, mNewEventColor);
             mNewEventIconDrawable = a.getDrawable(R.styleable.WeekView_newEventIconResource);
             mNewEventId = a.getInt(R.styleable.WeekView_newEventId, mNewEventId);
+            mNewEventIdentifier =  (a.getString(R.styleable.WeekView_newEventId) != null)? a.getString(R.styleable.WeekView_newEventId) : mNewEventIdentifier;
```
Comment (jhoobergs) [inline library/src/main/java/com/alamkanak/weekview/WeekView.java]: I think you should also update the `mNewEventIdentifier`  when the int version is set. This can easily be solved when you use the setter `setNewEventId`. The attribute `int mNewEventId ` isn't needed anymore, you should just convert the int's to strings.

File: library/src/main/java/com/alamkanak/weekview/WeekView.java [hunk 5]
```diff
@@ -1067,7 +1069,7 @@ private void drawEvents(Calendar date, float startFromPixel, Canvas canvas) {
                         if(mEventRects.get(i).event.getStartTime().get(Calendar.HOUR_OF_DAY) < mMinTime)
                             topToUse = mHourHeight *  getPassedMinutesInDay(mMinTime, 0) / 60 + mCurrentOrigin.y + mHeaderHeight + mHeaderRowPadding * 2 + mHeaderMarginBottom + mTimeTextHeight/2 + mEventMarginVertical - marginTop;
 
-                        if(mEventRects.get(i).event.getId() != mNewEventId)
+                        if(mEventRects.get(i).event.getIdentifier() != mNewEventIdentifier)
```
Comment (jhoobergs) [inline library/src/main/java/com/alamkanak/weekview/WeekView.java]: String comparison should happen with `.equals()`

File: library/src/main/java/com/alamkanak/weekview/WeekView.java [hunk 6]
```diff
@@ -1163,7 +1165,7 @@ private void drawEventTitle(WeekViewEvent event, RectF rect, Canvas canvas, floa
                 int availableLineCount = availableHeight / lineHeight;
                 do {
                     // Ellipsize text to fit into event rect.
-                    if (event.getId() != mNewEventId)
+                    if (event.getIdentifier() != mNewEventIdentifier)
```
Comment (jhoobergs) [inline library/src/main/java/com/alamkanak/weekview/WeekView.java]: String comparison

File: library/src/main/java/com/alamkanak/weekview/WeekViewEvent.java [hunk 7]
```diff
@@ -65,13 +65,49 @@ public WeekViewEvent(long id, String name, int startYear, int startMonth, int st
...
+     * @param endYear Year when the event ends.
+     * @param endMonth Month when the event ends.
+     * @param endDay Day when the event ends.
+     * @param endHour Hour (in 24-hour format) when the event ends.
+     * @param endMinute Minute when the event ends.
+     */
+    @Deprecated
+    public WeekViewEvent(long id, String name, int startYear, int startMonth, int startDay, int startHour, int startMinute, int endYear, int endMonth, int endDay, int endHour, int endMinute) {
+        this.mId = Integer.toString((int) id);
+
```
Comment (jhoobergs) [inline library/src/main/java/com/alamkanak/weekview/WeekViewEvent.java]: you should call the constructor with the String attribute after converting instead of copy pasting.

File: library/src/main/java/com/alamkanak/weekview/WeekViewEvent.java [hunk 8]
```diff
@@ -88,9 +124,30 @@ public WeekViewEvent(long id, String name, String location, Calendar startTime,
      * @param location The location of the event.
      * @param startTime The time when the event starts.
      * @param endTime The time when the event ends.
+     * @param allDay Is the event an all day event.
+     * @param shader the Shader of the event rectangle
+     */
+    @Deprecated
+    public WeekViewEvent(long id, String name, String location, Calendar startTime, Calendar endTime, boolean allDay, Shader shader) {
+        this.mId = Integer.toString((int) id);
```
Comment (jhoobergs) [inline library/src/main/java/com/alamkanak/weekview/WeekViewEvent.java]: Should also call the other constructor after conversion to string

File: library/src/main/java/com/alamkanak/weekview/WeekViewEvent.java [hunk 9]
```diff
@@ -101,22 +158,60 @@ public WeekViewEvent(long id, String name, String location, Calendar startTime,
      * @param location The location of the event.
      * @param startTime The time when the event starts.
      * @param endTime The time when the event ends.
+     * @param allDay Is the event an all day event
      */
-    public WeekViewEvent(long id, String name, String location, Calendar startTime, Calendar endTime) {
+    @Deprecated
+    public WeekViewEvent(long id, String name, String location, Calendar startTime, Calendar endTime, boolean allDay) {
+        this(Integer.toString((int) id), name, location, startTime, endTime, allDay, null);
```
Comment (jhoobergs) [inline library/src/main/java/com/alamkanak/weekview/WeekViewEvent.java]: This one is ok, but doesn't actually need to convert it to a string, the conversion would happen later the other constructor.

File: library/src/main/java/com/alamkanak/weekview/WeekViewEvent.java [hunk 10]
```diff
@@ -101,22 +158,60 @@ public WeekViewEvent(long id, String name, String location, Calendar startTime,
...
+    /**
+     * Initializes the event for week view.
+     * @param id The id of the event.
+     * @param name Name of the event.
+     * @param startTime The time when the event starts.
+     * @param endTime The time when the event ends.
+     */
+    @Deprecated
+    public WeekViewEvent(long id, String name, Calendar startTime, Calendar endTime) {
+        this(Integer.toString((int) id), name, null, startTime, endTime);
```
Comment (jhoobergs) [inline library/src/main/java/com/alamkanak/weekview/WeekViewEvent.java]: Could also call `public WeekViewEvent(long id, String name, String location, Calendar startTime, Calendar endTime, boolean allDay)`, the conversion to int would happen later

File: library/src/main/java/com/alamkanak/weekview/WeekViewEvent.java [hunk 11]
```diff
@@ -174,14 +269,24 @@ public void setShader(Shader shader){
...
+        return Integer.parseInt(mId);
+    }
+
+    public void setIdentifier(String id) {
         this.mId = id;
     }
 
+    @Deprecated
+    public void setId(long id) {
+        this.mId = Integer.toString((int) id);
```
Comment (jhoobergs) [inline library/src/main/java/com/alamkanak/weekview/WeekViewEvent.java]: The conversion to int can cause problems: use `String.valueOf(id)`

**Review by jhoobergs — CHANGES_REQUESTED** [review]
@udaysrinath Almost there 😄

File: library/src/main/java/com/alamkanak/weekview/WeekViewEvent.java (hunk 7 above)
Comment (jhoobergs) [reply library/src/main/java/com/alamkanak/weekview/WeekViewEvent.java]: The conversion should happen with `String.valueOf(id);`

**Comment by udaysrinath** [comment]
@jhoobergs - Addressed the review comments, mind taking another look?

File: library/src/main/java/com/alamkanak/weekview/WeekViewEvent.java [hunk 12]
```diff
@@ -174,14 +247,24 @@ public void setShader(Shader shader){
...
 
-    public long getId() {
+    public String getIdentifier() {
         return mId;
     }
 
-    public void setId(long id) {
+    @Deprecated
+    public long getId() {
+        return Integer.parseInt(mId);
```
Comment (jhoobergs) [inline library/src/main/java/com/alamkanak/weekview/WeekViewEvent.java]: `Long.parseLong`

**Review by jhoobergs — CHANGES_REQUESTED** [review]
Looks good.
Just one remark that I didn't see before

**Comment by udaysrinath** [comment]
@jhoobergs - Updated PR and squashed commits into 1 commit. Can you merge and publish a new version, would like to consume it and ensure all is great! Thanks again for the review and your help.

**Comment by jhoobergs** [comment]
@udaysrinath There is one thing I forgot (I'm sorry 😄 ) Can you also update the readme ? 
I'll test the code now

**Comment by udaysrinath** [comment]
Updated Readme, new version ```2.0.1```

**Comment by udaysrinath** [comment]
@jhoobergs -Just noticed this, previously the gradle dependency was ``` compile 'com.github.quivr:android-week-view:2.0.2' ```, even though the readme was ```2.0.0``` can you explain that? Is there a discrepancy here? Is the usage section a bug that indicated ```2.0.2```?

**Comment by jhoobergs** [comment]
@udaysrinath The changelog wasn't up to date, some bugfixes.
I pushed a new version called 2.2.0
Important: Automatic prefetching is disabled now, as described in the changelog

**Comment by udaysrinath** [comment]
Hmm, cant seem to find anything with ```2.2.0``` if I add that as a dependency.

**Comment by jhoobergs** [comment]
@udaysrinath Sometimes it takes some time. You can check http://search.maven.org/#search%7Cgav%7C1%7Cg%3A%22com.github.quivr%22%20AND%20a%3A%22android-week-view%22 to see if it should be available.
